vapi/appliance/simulator/simulator.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/simulator"
+	"github.com/vmware/govmomi/vapi/appliance/access/consolecli"
+	"github.com/vmware/govmomi/vapi/appliance/access/dcui"
+	"github.com/vmware/govmomi/vapi/appliance/access/shell"
+	"github.com/vmware/govmomi/vapi/appliance/access/ssh"
+	vapi "github.com/vmware/govmomi/vapi/simulator"
+)
+
+func init() {
+	simulator.RegisterEndpoint(func(s *simulator.Service, r *simulator.Registry) {
+		New(s.Listen).Register(s, r)
+	})
+}
+
+// Handler implements the Appliance API simulator
+type Handler struct {
+	sync.Mutex
+	URL              *url.URL
+	consoleCLIAccess consolecli.Access
+	dcuiAccess       dcui.Access
+	sshAccess        ssh.Access
+	shellAccess      shell.Access
+}
+
+// New creates a Handler instance
+func New(u *url.URL) *Handler {
+	return &Handler{
+		URL:              u,
+		consoleCLIAccess: consolecli.Access{Enabled: false},
+		dcuiAccess:       dcui.Access{Enabled: false},
+		sshAccess:        ssh.Access{Enabled: false},
+		shellAccess: shell.Access{
+			Enabled: false,
+			Timeout: 0,
+		},
+	}
+}
+
+// Register Appliance Management API paths with the vapi simulator's http.ServeMux
+func (h *Handler) Register(s *simulator.Service, r *simulator.Registry) {
+	s.HandleFunc(consolecli.Path, h.consoleCLI)
+	s.HandleFunc(dcui.Path, h.dcui)
+	s.HandleFunc(ssh.Path, h.ssh)
+	s.HandleFunc(shell.Path, h.shell)
+}
+
+func (h *Handler) decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	return Decode(r, w, val)
+}
+
+// Decode the request Body into val.
+// Returns true on success, otherwise false and sends the http.StatusBadRequest response.
+func Decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
```

Comment:
IMHO these should all be `500` in case of error according to the API spec?

update: just noted that apparently you did not update the doc string as it seems the handler implementations do this

Reply:
Where is the status send?

Reply:
Update the DOC string and code
```
// Decode the request Body into val.
// Returns true on success, otherwise false and sends the http.StatusInternalServerError response.
func Decode(request *http.Request, writer http.ResponseWriter, val interface{}) bool {
	defer request.Body.Close()
	err := json.NewDecoder(request.Body).Decode(val)
	if err != nil {
		log.Printf("%s %s: %s", request.Method, request.RequestURI, err)
		writer.WriteHeader(http.StatusInternalServerError)
		return false
	}
	return true
}
```